feat: Support OAuth Applications

AlexNti opened the pull request:
In this commit we are adding the CRUD methods and the rotate client secret for  for OAuth Applications, the documentation for those API can be found at https://clerk.com/docs/reference/backend-api/tag/OAuth-Applications

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	}
+}
+
+// fetches a single OAuth application by its ID.
```
```suggestion
// Get fetches a single OAuth application by its ID.
```

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	clerk.ListParams
+}
+
+func (params *ListParams) ToQuery() url.Values {
```
:question: Is this method needed? Since we're not adding functionality over clerk.ListParams.ToQuery it feels redundant.

AlexNti replied:
Without I am getting the following error, 

`cannot use params (variable of type *ListParams) as clerk.Params value in argument to req.SetParams: *ListParams does not implement clerk.Params (ambiguous selector *ListParams.ToQuery`

Is there any other way to bypass that error?

gkats replied:
I see, the problem is that both ListParams and APIParams implement the ToQuery method. This reveals a bigger problem with our types.

I'd suggest leaving as is for now. :smile:

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	return list, err
+}
+
+type OAuthApplicationCreateParams struct {
```
:wrench: 
```suggestion
type CreateParams struct {
```

Also, let's make all fields pointers to types (`*string` and `*bool`).

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// Creates a new OAuth application with the given parameters.
+func (c *Client) Create(ctx context.Context, params *OAuthApplicationCreateParams) (*clerk.OAuthApplicationWithClientSecret, error) {
```
:question: Do we need the different type here? Why can't `Create` return a `*clerk.OAuthApplication`?

```suggestion
func (c *Client) Create(ctx context.Context, params *CreateParams) (*clerk.OAuthApplication, error) {
```

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	Public      bool   `json:"public"`
+}
+
+// Creates a new OAuth application with the given parameters.
```
```suggestion
// Create creates a new OAuth application with the given parameters.
```

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	return authApplication, err
+}
+
+type OAuthApplicationUpdateParams struct {
```
```suggestion
type UpdateParams struct {
```

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+	Scopes      *string `json:"scopes,omitempty"`
+}
+
+// Updates an existing OAuth application.
```
```suggestion
// Update updates an existing OAuth application.
```

gkats commented on oauthapplication/client.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// Rotates the OAuth application's client secret
+func (c *Client) RotateClientSecret(ctx context.Context, id string) (*clerk.OAuthApplicationWithClientSecret, error) {
```
Let's use one type for `OAuthApplication`s.
```suggestion
func (c *Client) RotateClientSecret(ctx context.Context, id string) (*clerk.OAuthApplication, error) {
```

gkats commented on oauth_application.go:
```diff
@@ -0,0 +1,30 @@
+	UpdatedAt     int64  `json:"updated_at"`
+}
+
+type OAuthApplicationWithClientSecret struct {
```
:question: Do we need an additional type for just one field? How about we add a `*string` field to the `OAuthApplication` type instead?

```go
type OAuthApplication struct {
  ClientSecret *string `json:"client_secret,omitempty"`
}
```

AlexNti replied:
You are right this a better approach, I will update the code!

gkats reviewed (commented):
Nice one, @AlexNti!

I've left some comments, let me know what you think.

gkats reviewed (approved):
Before you merge this one, could you please also update the [CHANGELOG](https://github.com/clerk/clerk-sdk-go/blob/fe4b64f7cb6df287b94c226fae63faf271483895/CHANGELOG.md)?

You can add a "Next release" with your changes for now. See [this PR](https://github.com/clerk/clerk-sdk-go/pull/350/files) for an example.

AlexNti committed:
feat: Add support for OAuth applications

In this commit we are adding the CRUD methods and the rotate client secret for  for OAuth Applications, the documentation for those API can be found at https://clerk.com/docs/reference/backend-api/tag/OAuth-Applications
